File: python/ops_gnn/gather_coo.py
# Copyright (c) 2026 Huawei Technologies Co., Ltd.
# This program is free software, you can redistribute it and/or modify it
# under the terms and conditions of CANN Open Software License Agreement
# Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except
# in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY
# KIND, EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
# NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text.

"""Public Gather COO API."""

from typing import Optional

from . import _pybind
from .typing import Tensor


def gather_coo(
    src: Tensor,
    index: Tensor,
    out: Optional[Tensor] = None,
) -> Tensor:
    """Gather rows from ``src`` according to a COO-style index tensor.

    ``index`` has the same prefix shape as ``src`` and uses its last
    dimension as the entry dimension.  The corresponding dimension of
    ``src`` is replaced by ``index.size(-1)``; all suffix dimensions are
    copied unchanged.  Indices must be int64, sorted, and within the source
    row range as required by the task contract.  The operation is a raw
    element-wise copy and does not provide an autograd implementation.

    Args:
        src: NPU tensor with rank 1 through 8.
        index: NPU int64 tensor.  Its rank is at most ``src.dim()`` and its
            prefix dimensions must match ``src``.
        out: Optional destination tensor with the exact output shape and
            matching dtype/device.  ``None`` is distinct from a valid empty
            tensor and is forwarded explicitly to the native binding.

    Returns:
        ``out`` when supplied, otherwise a newly allocated NPU tensor.
    """
    return _pybind.gather_coo(src, index, out)

File: csrc/npu/gather_coo/op_kernel/arch35/gather_coo.cpp
/**
 * Copyright (c) 2026 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it
 * under the terms and conditions of CANN Open Software License Agreement
 * Version 2.0.
 * Please refer to the License for details.
 */

#include "gather_coo_kernel.h"
#include "gather_coo.h"

template <typename T>
void GatherCoo(
    T* src,
    int64_t* index,
    T* out,
    const GatherCooTilingData& tiling,
    aclrtStream stream)
{
    gather_coo_kernel<T><<<tiling.coreNum, nullptr, stream>>>(
        reinterpret_cast<uint8_t*>(src),
        reinterpret_cast<uint8_t*>(index),
        reinterpret_cast<uint8_t*>(out),
        tiling);
}

template void GatherCoo<uint8_t>(
    uint8_t*, int64_t*, uint8_t*, const GatherCooTilingData&, aclrtStream);
template void GatherCoo<uint16_t>(
    uint16_t*, int64_t*, uint16_t*, const GatherCooTilingData&, aclrtStream);
template void GatherCoo<uint32_t>(
    uint32_t*, int64_t*, uint32_t*, const GatherCooTilingData&, aclrtStream);
